dynamic/internal/shim/loader.go

```diff
@@ -0,0 +1,263 @@
+	"path/filepath"
+	"strings"
+
+	"github.com/apparentlymart/go-versions/versions"
+	plugin "github.com/hashicorp/go-plugin"
+	"github.com/hashicorp/terraform-plugin-go/tfprotov5"
+	"github.com/hashicorp/terraform-plugin-go/tfprotov6"
+	"github.com/hashicorp/terraform-plugin-mux/tf5to6server"
+	disco "github.com/hashicorp/terraform-svchost/disco"
+	"github.com/opentofu/opentofu/internal/getproviders"
+	"github.com/opentofu/opentofu/internal/logging"
+	tfplugin "github.com/opentofu/opentofu/internal/plugin"
+	"github.com/opentofu/opentofu/internal/providercache"
+	"github.com/opentofu/opentofu/internal/tfplugin5"
+	"github.com/opentofu/opentofu/internal/tfplugin6"
+	v5shim "github.com/opentofu/opentofu/shim/protov5"
+	v6shim "github.com/opentofu/opentofu/shim/protov6"
+	tfaddr "github.com/opentofu/registry-address"
+
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/contract"
+)
+
+// TODO:
+//
+// - If TF is setup and has already cached a provider, we should try to use that
+// provider.
+//
+// - If not, but TF is setup we should probably contribute to the existing cache.
+//
+// - If TF is not setup, we should cache within PULUMI_HOME to avoid creating new dirs.
+const envPluginCache = "TF_PLUGIN_CACHE_DIR"
+
+// A loaded and running provider.
+//
+// You must call Close on any Provider that has been created.
+type Provider interface {
+	tfprotov6.ProviderServer
+	io.Closer
+
+	Name() string
+	Version() string
+}
+
+// Load a TF provider with key and version specified.
+//
+// If version is "", then whatever version is currently installed will be used. If no
+// version is installed then the latest version can be used.
+//
+// `=`, `<=`, `>=` sigils can be used just like in TF.
+func LoadProvider(ctx context.Context, key, version string) (Provider, error) {
+
+	p, err := tfaddr.ParseProviderSource(key)
+	if err != nil {
+		return nil, fmt.Errorf("invalid provider name: %w", err)
+	}
+
+	v, err := getproviders.ParseVersionConstraints(version)
+	if err != nil {
+		return nil, err
+	}
+
+	return getProviderServer(ctx, p, v)
+}
+
+// RunLocalProvider runs a provider by it's path.
+func RunLocalProvider(ctx context.Context, path string) (Provider, error) {
+	dir, name, ok := cutLast(path, "terraform-provider-")
+	if !ok {
+		return nil, fmt.Errorf("expected path to end with %q", "terraform-provider-${NAME}")
+	}
+
+	return runProvider(ctx, &providercache.CachedProvider{
+		Provider:   tfaddr.Provider{Type: name},
+		Version:    versions.Version{},
+		PackageDir: dir,
+	})
+}
+
+func cutLast(s, sep string) (string, string, bool) {
+	if i := strings.LastIndex(s, sep); i >= 0 {
+		return s[:i], s[i+len(sep):], true
+	}
+
+	return s, "", false
+}
+
+func getPluginCache() (string, error) {
+	if dir := os.Getenv(envPluginCache); dir != "" {
+		return dir, nil
+	}
+
+	pulumiHome := os.Getenv("PULUMI_HOME")
+	if pulumiHome == "" {
+		home, err := os.UserHomeDir()
+		if err != nil {
+			return "", err
+		}
+		pulumiHome = filepath.Join(home, ".pulumi")
```

Comment:
pulumi/pulumi may have some linkable helper code to resolve pulumi home.

Reply:
It does:

https://github.com/pulumi/pulumi/blob/7c0489fafcdc5c393f06ced3e0e8b2da413bcbec/sdk/go/common/workspace/paths.go#L264-L266

Thanks for suggesting it.